Index: gobject/valaccodedynamicmethodbinding.vala
===================================================================
--- gobject/valaccodedynamicmethodbinding.vala	(revision 1337)
+++ gobject/valaccodedynamicmethodbinding.vala	(working copy)
@@ -234,18 +234,23 @@
 
 				block.add_statement (new CCodeExpressionStatement (ccall));
 
-				block.add_statement (new CCodeReturnStatement (new CCodeIdentifier ("result")));
 
 				if (array_type != null) {
 					// special case string array
 					if (!method.no_array_length) {
 						var cstrvlen = new CCodeFunctionCall (new CCodeIdentifier ("g_strv_length"));
 						cstrvlen.add_argument (new CCodeIdentifier ("result"));
-						expr.append_array_size (cstrvlen);
+						var ass = new CCodeAssignment(new 
+									CCodeUnaryExpression(CCodeUnaryOperator.POINTER_INDIRECTION, new 
+									CCodeIdentifier(codegen.get_array_length_cname("result", 1))), 
+									cstrvlen);
+						block.add_statement (new CCodeExpressionStatement (ass));
+						//expr.append_array_size(new CCodeIdentifier(codegen.get_array_length_cname("result",1)));
 					} else {
-						expr.append_array_size (new CCodeConstant ("-1"));
+						//expr.append_array_size (new CCodeConstant ("-1"));
 					}
 				}
+				block.add_statement (new CCodeReturnStatement (new CCodeIdentifier ("result")));
 			}
 		} else {
 			block.add_statement (new CCodeExpressionStatement (ccall));
