Index: vapi/glib-2.0.vapi
===================================================================
--- vapi/glib-2.0.vapi	(revision 1349)
+++ vapi/glib-2.0.vapi	(working copy)
@@ -514,6 +514,8 @@
 	public string escape (string exceptions);
 	[CCode (cname = "g_strcompress")]
 	public string compress ();
+	[CCode (cname = "g_strdelimit")]
+	public weak string delimit (string delimiters, char delimiter);
 	[CCode (cname = "g_strsplit")]
 	[NoArrayLength]
 	public string[] split (string delimiter, int max_tokens = 0);
