Index: glib-2.0.vapi
===================================================================
--- vapi/glib-2.0.vapi	(revision 1345)
+++ vapi/glib-2.0.vapi	(working copy)
@@ -514,6 +514,8 @@
 	public string escape (string exceptions);
 	[CCode (cname = "g_strcompress")]
 	public string compress ();
+	[CCode (cname = "g_strdelimit")]
+	public weak string delimit (string delimiters, char delimiter);
 	[CCode (cname = "g_strsplit")]
 	[NoArrayLength]
 	public string[] split (string delimiter, int max_tokens = 0);
@@ -525,6 +527,9 @@
 	public static string joinv (string separator, string[] str_array);
 	[CCode (cname = "g_strnfill")]
 	public static string nfill (ulong length, char fill_char);
+	[CCode (cname = "g_strv_length")]
+	[NoArrayLength]
+	public static int lengthv (string[] str_array);
 
 	[CCode (cname = "g_utf8_next_char")]
 	public weak string next_char ();
