Index: vapi/glib-2.0.vapi
===================================================================
--- vapi/glib-2.0.vapi	(revision 1325)
+++ vapi/glib-2.0.vapi	(working copy)
@@ -525,6 +525,9 @@
 	public static string joinv (string separator, string[] str_array);
 	[CCode (cname = "g_strnfill")]
 	public static string nfill (ulong length, char fill_char);
+	[CCode (cname = "g_strv_length")]
+	[NoArrayLength]
+	public static int lengthv (string[] str_array);
 
 	[CCode (cname = "g_utf8_next_char")]
 	public weak string next_char ();
