Index: gtk+-2.0.vapi
===================================================================
--- vapi/gtk+-2.0.vapi	(révision 1188)
+++ vapi/gtk+-2.0.vapi	(copie de travail)
@@ -1305,7 +1305,6 @@
 	}
 	[CCode (cheader_filename = "gtk/gtk.h")]
 	public class NotebookPage {
-		public static int num (Gtk.Notebook notebook, Gtk.Widget child);
 	}
 	[CCode (cheader_filename = "gtk/gtk.h")]
 	public class PageRange {
@@ -3636,6 +3635,7 @@
 		public int insert_page_menu (Gtk.Widget child, Gtk.Widget tab_label, Gtk.Widget menu_label, int position);
 		public Notebook ();
 		public void next_page ();
+		public int page_num (Gtk.Widget child);
 		public void popup_disable ();
 		public void popup_enable ();
 		public int prepend_page (Gtk.Widget child, Gtk.Widget tab_label);

