--- gtk2.spec	2007-10-25 09:04:49.000000000 -0400
+++ gtk2-aqd.spec	2007-12-07 23:47:58.000000000 -0500
@@ -14,7 +14,7 @@
 %define bin_version 2.10.0
 
 Summary: The GIMP ToolKit (GTK+), a library for creating GUIs for X
-Name: gtk2
+Name: gtk2-aqd
 Version: %{base_version}
 Release: 5%{?dist}
 License: LGPLv2+
@@ -22,7 +22,9 @@
 Source: http://download.gnome.org/sources/gtk+/2.12/gtk+-%{version}.tar.bz2
 Source1: update-gdk-pixbuf-loaders
 Source2: update-gtk-immodules 
-
+Source9: http://aur.archlinux.org/packages/gtk2-aqd/gtk2-aqd.tar.gz
+Provides: gtk2 = %{base_version}-%{release}
+Obsoletes: gtk2 = %{base_version}-%{release}
 # Biarch changes
 Patch0: gtk+-2.4.1-lib64.patch
 # Fedora patch
@@ -91,7 +93,7 @@
 %package devel
 Summary: Development tools for GTK+ applications
 Group: Development/Libraries
-Requires: gtk2 = %{version}-%{release}
+Requires: gtk2-aqd = %{version}-%{release}
 Requires: pango-devel >= %{pango_version}
 Requires: atk-devel >= %{atk_version}
 Requires: glib2-devel >= %{glib2_version}
@@ -105,7 +107,8 @@
 Requires: automake
 # for /usr/share/gtk-doc/html
 Requires: gtk-doc
-
+Provides: gtk2-devel = %{version}-%{release}
+Obsoletes: gtk2-devel = %{version}-%{release}
 %description devel
 The gtk+-devel package contains the header files and developer
 docs for the GTK+ widget toolkit.  
@@ -124,6 +127,11 @@
   test -f %{_datadir}/libtool/$i && cp %{_datadir}/libtool/$i .
 done
 
+tar --strip-components 1 -xzvf $RPM_SOURCE_DIR/gtk2-aqd.tar.gz -C $RPM_BUILD_DIR/gtk+-%{version}
+patch -p1 < gdkscreen-check.patch
+#patch -p1 < gtkclipboard-check.patch
+patch -p0 < gtkmenubar.patch
+
 %build
 libtoolize --force
 
